packages/instruction-plans/src/transaction-planner-base.ts

```diff
@@ -0,0 +1,416 @@
+            transactionPlans.push(newPlan);
+        }
+    }
+
+    if (transactionPlans.length === 1) {
+        return transactionPlans[0];
+    }
+    if (transactionPlans.length === 0) {
+        return null;
+    }
+    if (context.parent?.kind === 'parallel') {
+        return { kind: 'parallel', plans: transactionPlans };
+    }
+    return {
+        divisible: context.parent?.kind === 'sequential' ? context.parent.divisible : true,
+        kind: 'sequential',
+        plans: transactionPlans,
+    };
+}
+
+function getSequentialCandidate(latestPlan: MutableTransactionPlan): MutableSingleTransactionPlan | null {
+    if (latestPlan.kind === 'single') {
+        return latestPlan;
+    }
+    if (latestPlan.kind === 'sequential' && latestPlan.plans.length > 0) {
+        return getSequentialCandidate(latestPlan.plans[latestPlan.plans.length - 1]);
+    }
+    return null;
+}
+
+function getParallelCandidates(latestPlan: TransactionPlan): MutableSingleTransactionPlan[] {
+    return getAllSingleTransactionPlans(latestPlan);
+}
+
+async function selectAndMutateCandidate(
+    context: Pick<TraverseContext, 'abortSignal' | 'onTransactionMessageUpdated'>,
+    candidates: MutableSingleTransactionPlan[],
+    predicate: (
+        message: BaseTransactionMessage & TransactionMessageWithFeePayer,
+    ) => BaseTransactionMessage & TransactionMessageWithFeePayer,
+): Promise<MutableSingleTransactionPlan | null> {
+    for (const candidate of candidates) {
+        try {
+            const message = await context.onTransactionMessageUpdated(predicate(candidate.message), {
+                abortSignal: context.abortSignal,
+            });
+            if (getTransactionMessageSize(message) <= TRANSACTION_SIZE_LIMIT) {
+                candidate.message = message;
+                return candidate;
+            }
+        } catch (error) {
+            if (
+                isSolanaError(error, SOLANA_ERROR__INSTRUCTION_PLANS__CANNOT_PACK_USING_PROVIDED_MESSAGE) ||
+                isSolanaError(error, SOLANA_ERROR__INSTRUCTION_PLANS__CANNOT_FIT_ENTIRE_INSTRUCTION_PLAN_INSIDE_MESSAGE)
+            ) {
+                // Try the next candidate.
+            } else {
+                throw error;
+            }
+        }
+    }
+    return null;
+}
+
+async function createNewMessage(
+    context: Pick<TraverseContext, 'abortSignal' | 'createTransactionMessage' | 'onTransactionMessageUpdated'>,
+    instructionPlan: InstructionPlan,
+    predicate: (
+        message: BaseTransactionMessage & TransactionMessageWithFeePayer,
+    ) => BaseTransactionMessage & TransactionMessageWithFeePayer,
+): Promise<BaseTransactionMessage & TransactionMessageWithFeePayer> {
+    const newMessage = await context.createTransactionMessage({
+        abortSignal: context.abortSignal,
+    });
+    const updatedMessage = await context.onTransactionMessageUpdated(predicate(newMessage), {
+        abortSignal: context.abortSignal,
+    });
```

Comment:
Same here. If the user-supplied functions punk out and don't bother to implement abort-signal-listening code then firing the signal will do nothing.